### PR TITLE
Removes ConcurrencyEnforcer—and by extension also Warmup

### PR DESCRIPTION
**Pull request by viktorklang**
See #413 

Ping @jroper & @pvlugter regarding if we should make an effort to keep Warmup or not, given graalvm native-image

### CODE REVIEW

**Comment by ralphlaude** [comment]
@viktorklang nice to remove code! could i know why it is not longer necessary (it is not explained here #413)?

**Review by pvlugter — APPROVED** [review]
LGTM

It would be interesting to run load tests with the concurrency enforcer removed. I wonder if it will be easy to end up with some over-parallelisation slowdowns, depending on the concurrency settings in the proxy and user function.

I think it's fine to remove the warmup as it is now. There's the native-images. And for running with hotspot JVMs, using CDS (class data sharing to load the classes from an image) and jaotc (ahead-of-time compile up to tier 2), and then have the JIT compiler continue from there, could be the best way to improve startup.

**Comment by pvlugter** [comment]
One of the go operator tests may be flaky, or just need more time to wait for a check. I'll restart that test.

**Comment by sleipnir** [comment]
> LGTM
> 
> It would be interesting to run load tests with the concurrency enforcer removed. I wonder if it will be easy to end up with some over-parallelisation slowdowns, depending on the concurrency settings in the proxy and user function.
> 
> I think it's fine to remove the warmup as it is now. There's the native-images. And for running with hotspot JVMs, using CDS (class data sharing to load the classes from an image) and jaotc (ahead-of-time compile up to tier 2), and then have the JIT compiler continue from there, could be the best way to improve startup.

@pvlugter  using Openjdk / openJ9 using xshareClasses is a great option, I have used it in production with great results on many of our JVM based systems.

**Comment by jroper** [comment]
Warmup still has advantages, there's a lot of things that get initialised and warmed during warm up beyond JIT, the biggest thing being database connection pools.

File: proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/EventSourcedEntity.scala [hunk 1]
```diff
@@ -253,9 +244,8 @@ final class EventSourcedEntity(configuration: EventSourcedEntity.Configuration,
     currentCommand =
       EventSourcedEntity.OutstandingCommand(idCounter, actorId + ":" + entityId + ":" + idCounter, sender)
     commandStartTime = System.nanoTime()
-    concurrencyEnforcer ! Action(currentCommand.actionId, () => {
-      relay ! EventSourcedStreamIn(EventSourcedStreamIn.Message.Command(command))
-    })
+    relay ! EventSourcedStreamIn(EventSourcedStreamIn.Message.Command(command))
+    statsCollector ! StatsCollector.CommandSent
```
Comment (jroper) [inline proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/EventSourcedEntity.scala:248]: I think you can remove stats collector too... it's only used for autoscaling.

**Comment by pvlugter** [comment]
> Warmup still has advantages, there's a lot of things that get initialised and warmed during warm up beyond JIT, the biggest thing being database connection pools.

The warmup looks like it just processes one command. Doesn't seem like that would have any affect on JIT. So it's more about class loading and initialising a database connection as early as possible?

**Comment by jroper** [comment]
Yeah I should have said class loading not JIT. There's probably also a significant amount of other initialisation, eg reflection etc, as various components are loaded and initialised for the first time.

File: proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/EventSourcedEntity.scala (hunk 1 above)
Comment (viktorklang) [reply proxy/core/src/main/scala/io/cloudstate/proxy/eventsourced/EventSourcedEntity.scala:248]: @jroper Don't we want autoscaling? :)

**Comment by viktorklang** [comment]
Let's see what happens with the native image bahavior when the warmup is gone. If we need to we'll reintrocue something which achieves the same effect, but extrapolated over more than just the ES code.

**Pull request merged**